app/Model/Users.php

```
@@ -74,51 +76,49 @@ public function loadUser(User $user)
 
         $query = "select id, email, password, verifyCode, passwordResetCode, withdrawalAddress from users where $where limit 1";
         $rows = $this->mysql->query($query, $types, $params);
-        
-        if (empty($rows) || !isset($rows[0])) {
-            throw new \Exception('Unable to find user.');
+
+        if ($rows || isset($rows[0])) {
+            $user->init($rows[0]);
```

This is not the correct fix. This code should remain the same. The bug is caused because the LoginForm should use Validator::isValidEmailString instead of Validator::isValidEmailChars (and same for password).

---

It also fail. i made it isValidEmailString && isValidPasswordString
but still the error shows up.
The thing is if i'll use this email : momen@said.shaker 
and password : 123456 they will pass the isValidEmailString && isValidPasswordString 
and when they reach the isCorrectPassword condition, it should create a user and load the user
which will reach the exception.
you decide!!.

---

Oh, I see what you are saying. I think we are missing this code before loading:

```
        if (!$this->users->emailExists($email)) {
            return false;
        }
```
inside the LoginForm::isCorrectPassword() to make sure it exists before loading the email. Does it make sense?

Although your solution could possibly work too, the reason I choose checking the email first is because there are situations where loadUser() should always work such as loading a user in a session, and in those situations, we want to throw an Exception to indicate an unexpected problem where the user is not found. So the caller should know, before loading the user, if the email exists.

---

Yes, I see.